src/ape_run/_cli.py

```diff
@@ -1,33 +1,155 @@
+from types import ModuleType
+from typing import Dict, Union
+
 import click
 
-from ape.cli import NetworkBoundCommand, ape_cli_context, network_option
+from ape.logging import logger
+from ape.managers.project import ProjectManager
+from ape.utils import cached_property, get_relative_path
+from ape_console._cli import console
+
+
+def deepcopy_skip_modules(d):
+    return {
+        k: deepcopy_skip_modules(v) if isinstance(v, dict) else v
+        for k, v in d.items()
+        if not k.startswith("_") and not isinstance(v, ModuleType)
+    }
+
+
+class ScriptCommand(click.MultiCommand):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs, result_callback=self.result_callback)
+        self.ns: dict = {}
+
+    def __get_command(self, filepath: Path) -> Union[click.Command, click.Group, None]:
+        relative_filepath = get_relative_path(filepath, self._project.path)
+
+        # First load the code module by compiling it
+        # NOTE: This does not execute the module
+        logger.debug(f"Importing module: {relative_filepath}")
+        try:
+            code = compile(filepath.read_text(), filepath, "exec")
+        except SyntaxError as e:
+            logger.warning(f"Error parsing {relative_filepath}:\n\t{e.__class__.__name__}: {e}")
+            return None  # Prevents stalling scripts
+
+        # NOTE: Introspect code structure only for given patterns (do not execute it to find hooks)
+        if "cli" in code.co_names:
+            # If the module contains a click cli subcommand, process it and return the subcommand
+            logger.debug(f"Found 'cli' command in script: {relative_filepath}")
+
+            ns: dict = {}
+            try:
+                # TODO: Analyze security issues from this
+                eval(code, ns, ns)
+            except Exception as e:
+                logger.warning(f"Error loading {relative_filepath}:\n\t{e.__class__.__name__}: {e}")
+                return None  # NOTE: Allow other scripts to load if one breaks
+
+            # NOTE: So we can get the extra locals on callback
+            self.ns[filepath.stem] = deepcopy_skip_modules(ns)
+            return ns["cli"]  # retrun subcommand
+
+        elif "main" in code.co_names:
+            logger.debug(f"Found 'main' method in script: {relative_filepath}")
+
+            @click.command(short_help=f"Run '{relative_filepath}:main'")
+            def call():
+                ns: dict = {}
+                # TODO: Analyze security issues from this
+                eval(code, ns, ns)
+                # NOTE: So we can get the extra locals on callback
+                self.ns[filepath.stem] = deepcopy_skip_modules(ns)
+
+                ns["main"]()  # Call function
+
+            return call
+
+        else:
+            logger.warning(f"No 'main' method or 'cli' command in script: {relative_filepath}")
+
+            @click.command(short_help=f"Run '{relative_filepath}'")
+            def call():
```

Comment:
this call could be a default call above and the `elif` above could call it before doing `ns["main"]()  # Call function`